### PR TITLE
refactor api response pattern

### PR DESCRIPTION
**Pull request by gmrabian**
### Description

New api response pattern. In short:
```ts
// Rather than constructing a response object on the fly (current state):
if (!allParamsValid) {
  return {
    status: StatusCodes.BadRequest,
    body: error.NO_KEY,
  };
}

// We will construct a response with a named helper:
if (!allParamsValid) {
  return badRequest(error.NO_KEY);
}
```

Additionally we had certain return blocks based on variables being undefined where they could not be undefined as a result of their flow. This fixes that logic with try/catches using the new response pattern.

Review guidance: Review the response-lib changes. Pick a handler and read the changes. Notice the new response pattern. Notice any changes to try/catch blocks. These changes will be very similar in all the other handlers.

Moving the work from these two MFP PRs to MCR:
https://github.com/Enterprise-CMCS/macpro-mdct-mfp/pull/754
https://github.com/Enterprise-CMCS/macpro-mdct-mfp/pull/762

### Related ticket(s)

CMDCT-none

---
### How to test

This PR touches almost every file in the API, and most changes have to do with requests that are invalid, for one reason or another. That, together with AWS' request signing, makes it very difficult to test from the UI. But, the expected behavioral changes are also very small - for example, changing a 403 response to a 409, for a situation that we don't expect to ever happen.

For that reason, I would say that the automated e2e tests are sufficient here - coupled, of course, with a thorough reading of the code changes.

---
### Pre-review checklist

- [x] I have added [thorough](https://shorturl.at/aejkF) tests, if necessary
- ~[ ] I have updated relevant documentation, if necessary~
- [x] I have performed a self-review of my code
- [x] I have manually tested this PR in the deployed cloud environment

### CODE REVIEW

**Review by karla-vm — APPROVED** [review]
great work on this, Garrett!! 🥇

File: services/app-api/handlers/banners/create.ts [hunk 1]
```diff
@@ -5,51 +5,61 @@ import dynamoDb from "../../utils/dynamo/dynamodb-lib";
...
-      status: StatusCodes.UNAUTHORIZED,
-      body: error.UNAUTHORIZED,
-    };
-  } else if (!event?.pathParameters?.bannerId!) {
-    throw new Error(error.NO_KEY);
-  } else {
-    const unvalidatedPayload = JSON.parse(event!.body!);
+    return forbidden(error.UNAUTHORIZED);
+  }
+  if (!event?.pathParameters?.bannerId!) {
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/banners/create.ts:30]: Looks like you can do `!event?.pathParameters?.bannerId` and not suppress TS.

File: services/app-api/handlers/banners/create.ts [hunk 2]
```diff
@@ -5,51 +5,61 @@ import dynamoDb from "../../utils/dynamo/dynamodb-lib";
...
-          link: validatedPayload.link,
-          startDate: validatedPayload.startDate,
-          endDate: validatedPayload.endDate,
-        },
-      };
-      await dynamoDb.put(params);
-      return { status: StatusCodes.CREATED, body: params };
-    }
+  const params = {
+    TableName: process.env.BANNER_TABLE_NAME!,
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/banners/create.ts:46]: `process.env.BANNER_TABLE_NAME` passes.

File: services/app-api/handlers/banners/delete.ts [hunk 3]
```diff
@@ -3,25 +3,23 @@ import handler from "../handler-lib";
...
-      TableName: process.env.BANNER_TABLE_NAME!,
-      Key: {
-        key: event?.pathParameters?.bannerId!,
-      },
-    };
-    await dynamoDb.delete(params);
-    return { status: StatusCodes.SUCCESS, body: params };
+    return forbidden(error.UNAUTHORIZED);
   }
+  if (!event?.pathParameters?.bannerId!) {
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/banners/delete.ts:14]: `!event?.pathParameters?.bannerId`

File: services/app-api/handlers/banners/delete.ts [hunk 4]
```diff
@@ -3,25 +3,23 @@ import handler from "../handler-lib";
...
+    return badRequest(error.NO_KEY);
+  }
+  const params = {
+    TableName: process.env.BANNER_TABLE_NAME!,
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/banners/delete.ts:18]: `process.env.BANNER_TABLE_NAME`

File: services/app-api/handlers/reports/archive.ts [hunk 5]
```diff
@@ -4,62 +4,64 @@ import { fetchReport } from "./fetch";
...
   // Get current report
   const reportEvent = { ...event, body: "" };
   const getCurrentReport = await fetchReport(reportEvent, context);
 
-  if (!getCurrentReport.body) {
-    return {
-      status: StatusCodes.NOT_FOUND,
-      body: error.NO_MATCHING_RECORD,
-    };
+  if (!getCurrentReport?.body) {
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/archive.ts:36]: `getCurrentReport.body` seems to be okay.

File: services/app-api/handlers/reports/archive.ts [hunk 6]
```diff
@@ -4,62 +4,64 @@ import { fetchReport } from "./fetch";
...
-  // if current report exists, parse for archived status
-  if (getCurrentReport?.body) {
-    const currentReport = JSON.parse(getCurrentReport.body);
-    const currentArchivedStatus = currentReport?.archived;
-    const reportType = currentReport?.reportType;
-
-    const reportTable = reportTables[reportType as keyof typeof reportTables];
+  // parse for archived status
+  const currentReport = JSON.parse(getCurrentReport.body);
+  const currentArchivedStatus = currentReport?.archived;
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/archive.ts:42]: `currentReport.archived` and `currentReport.reportType` seem okay.

File: services/app-api/handlers/reports/create.ts [hunk 7]
```diff
@@ -100,17 +86,19 @@ export const createReport = handler(async (event, _context) => {
...
-      body: error.INVALID_DATA,
-    };
+  let validatedFieldData;
+  try {
+    validatedFieldData = await validateFieldData(
+      formTemplate.validationJson,
+      unvalidatedFieldData
+    );
+  } catch {
+    return badRequest(error.INVALID_DATA);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/create.ts:96]: Is this testable?

File: services/app-api/handlers/reports/create.ts [hunk 8]
```diff
@@ -145,25 +133,20 @@ export const createReport = handler(async (event, _context) => {
...
-      body: error.INVALID_DATA,
-    };
+  let validatedMetadata;
+  try {
+    validatedMetadata = await validateData(metadataValidationSchema, {
+      ...unvalidatedMetadata,
+    });
+  } catch {
+    // Return INVALID_DATA error if metadata is not valid.
+    return badRequest(error.INVALID_DATA);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/create.ts:146]: Is this testable?

File: services/app-api/handlers/reports/fetch.ts [hunk 9]
```diff
@@ -18,31 +18,28 @@ import {
...
   }
 
   const { reportType, state, id } = event.pathParameters!;
 
   if (!isState(state)) {
-    return {
-      status: StatusCodes.BAD_REQUEST,
-      body: error.NO_KEY,
-    };
+    return badRequest(error.NO_KEY);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/fetch.ts:39]: Is this testable?

File: services/app-api/handlers/reports/fetch.ts [hunk 10]
```diff
@@ -74,10 +68,7 @@ export const fetchReport = handler(async (event, _context) => {
 
     const formTemplate = (await s3Lib.get(formTemplateParams)) as AnyObject; // TODO: strict typing
     if (!formTemplate) {
-      return {
-        status: StatusCodes.NOT_FOUND,
-        body: error.MISSING_FORM_TEMPLATE,
-      };
+      return notFound(error.NO_MATCHING_RECORD);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/fetch.ts:71]: Is this testable?

File: services/app-api/handlers/reports/fetch.ts [hunk 11]
```diff
@@ -89,10 +80,7 @@ export const fetchReport = handler(async (event, _context) => {
     const fieldData = (await s3Lib.get(fieldDataParams)) as AnyObject; // TODO: strict typing
 
     if (!fieldData) {
-      return {
-        status: StatusCodes.NOT_FOUND,
-        body: error.NO_MATCHING_RECORD,
-      };
+      return notFound(error.NO_MATCHING_RECORD);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/fetch.ts:83]: Is this testable?

File: services/app-api/handlers/reports/submit.ts [hunk 12]
```diff
@@ -29,25 +37,16 @@ export const submitReport = handler(async (event, _context) => {
...
   }
 
   const { id, state, reportType } = event.pathParameters;
 
   if (!isState(state)) {
-    return {
-      status: StatusCodes.BAD_REQUEST,
-      body: error.NO_KEY,
-    };
+    return badRequest(error.NO_KEY);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/submit.ts:46]: Is this testable?

File: services/app-api/handlers/reports/submit.ts [hunk 13]
```diff
@@ -59,141 +58,108 @@ export const submitReport = handler(async (event, _context) => {
...
+    return notFound(error.NOT_IN_DATABASE);
+  }
+
+  const reportMetadata = response.Item as
+    | MLRReportMetadata
+    | MCPARReportMetadata;
+  const { status, isComplete, fieldDataId, formTemplateId } = reportMetadata;
+
+  if (status === "Submitted") {
+    return ok(reportMetadata);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/submit.ts:72]: Is this testable?

File: services/app-api/handlers/reports/update.ts [hunk 14]
```diff
@@ -147,6 +127,9 @@ export const updateReport = handler(async (event, context) => {
     string,
     any
   >;
+  if (!existingFieldData) {
+    return notFound(error.MISSING_DATA);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/update.ts:131]: Is this testable?

File: services/app-api/handlers/reports/update.ts [hunk 15]
```diff
@@ -155,31 +138,23 @@ export const updateReport = handler(async (event, context) => {
     unvalidatedPayload;
 
   if (!unvalidatedFieldData) {
-    return {
-      status: StatusCodes.BAD_REQUEST,
-      body: error.MISSING_DATA,
-    };
+    return badRequest(error.MISSING_DATA);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/update.ts:141]: Is this testable?

File: services/app-api/handlers/reports/update.ts [hunk 16]
```diff
@@ -155,31 +138,23 @@ export const updateReport = handler(async (event, context) => {
...
   }
 
   // Validation JSON should be there—if it's not, there's an issue.
   if (!formTemplate.validationJson) {
-    return {
-      status: StatusCodes.BAD_REQUEST,
-      body: error.MISSING_FORM_TEMPLATE,
-    };
+    return internalServerError(error.MISSING_FORM_TEMPLATE);
```
Comment (bangbay-bluetiger) [inline services/app-api/handlers/reports/update.ts:146]: Is this testable?

**Review by bangbay-bluetiger — APPROVED** [review]
Looks good! Non-blocking comments to address removing some `!` and `?`, and add more tests. Address if they haunt you. 👻

**Comment by gmrabian** [comment]
@bangbay-bluetiger I appreciate the suggestions but this PR is haunting me more than those comments 😅 

as to the tests I think some of those are testable but since I had so many lines to cover I went for the repeatable ones like dynamo and s3

**Review by benmartin-coforma — COMMENTED** [review]
Excellent work, and very much appreciated 🚀

**Pull request merged**